lib/jsonwp-proxy/proxy.js

```diff
@@ -232,15 +236,111 @@ class JWProxy {
     }
   }
 
+  async getDownstreamProtocol () {
+    if (!this._downstreamProtocol) {
+      const [, resBody] = await this.proxy('/status', 'GET');
+      this._downstreamProtocol = this.getProtocolFromResBody(resBody);
+    }
+    return this._downstreamProtocol;
+  }
+
+  requestToCommandName (url, method) {
+    const extractCommandName = (pattern) => {
+      const pathMatch = pattern.exec(url);
+      return pathMatch ? routeToCommandName(pathMatch[1], method) : null;
+    };
+    let commandName = routeToCommandName(url, method);
+    if (!commandName && _.includes(url, '/wd/hub/session/')) {
+      commandName = extractCommandName(/\/wd\/hub\/session\/[^\/]+(.+)/);
+    }
+    if (!commandName && _.includes(url, '/wd/hub/')) {
+      commandName = extractCommandName(/\/wd\/hub(\/.+)/);
+    }
+    return commandName;
+  }
+
+  async proxyCommand (url, method, body = null) {
+    const commandName = this.requestToCommandName(url, method);
+    if (!commandName) {
+      return await this.proxy(url, method, body);
+    }
+    log.debug(`Matched '${url}' to command name '${commandName}'`);
+
+    // Handle "crossing" endpoints for the case
+    // when upstream and downstream drivers operate different protocols
+
+    // Same url, but different arguments
+
+    if (commandName === 'timeouts')  {
+      return await this.proxySetTimeouts(url, method, body);
+    }
+
+    // Same arguments, but different URLs
+
+    const conflicts = [
+      {
+        commandNames: ['execute', 'executeAsync'],
+        jsonwpConverter: () => url.replace(/\/execute.*/,
+          url.includes('async') ? '/execute_async' : '/execute'),
+        w3cConverter: () => url.replace(/\/execute.*/,
+          url.includes('async') ? '/execute/async' : '/execute/sync'),
+      },
+      {
+        commandNames: ['getElementScreenshot'],
+        jsonwpConverter: () => url.replace(/\/element\/([^\/]+)\/screenshot$/,
+          '/screenshot/$1'),
+        w3cConverter: () => url.replace(/\/screenshot\/(.+)$/,
+          '/element/$1/screenshot'),
+      },
+    ];
+
+    for (const {commandNames, jsonwpConverter, w3cConverter} of conflicts) {
+      if (!commandNames.includes(commandName)) {
+        continue;
+      }
+
+      try {
```

Comment:
why do we try the proxy first not knowing whether it will work? could this conceivably have side effects? and wouldn't we only want to run the catch block in the case of a protocol mismatch error? how are we filtering out all other kinds of errors?

i think we could safely check the downstream protocol once for each session, and then just save it. then we simply cross protocols if we know we need to.